File: Python/Word Break.py
class Solution:
    # @param s, a string
    # @param dict, a set of string
    # @return a boolean
    def wordBreak(self, s, dict):
        len1 = len(s)
        seq = [0] * (len1 + 1)
        seq[0] = 1
        for i in range(1, len1+1):
            for j in range(i+1):
                if s[j:i] in dict and seq[j] == 1:
                    seq[i] = 1
                    break
        if seq[len1] == 1:
            return True
        else:
            return False

File: C++/Word-Break.cpp
class Solution {
public:
    bool wordBreak(string s, unordered_set<string>& wordDict) {
        bool *CanBreak = new bool[s.size() + 1];
        CanBreak[0] = true;
        for (int i = 1; i <= s.size(); i ++) {
            CanBreak[i] = false;
            for (int j = 0; j < i; j ++) {
                if (CanBreak[j] && wordDict.find(s.substr(j, i - j)) != wordDict.end()) {
                    CanBreak[i] = true;
                    break;
                }
            }
        }
        bool res = CanBreak[s.size()];
        delete []CanBreak;
        return res;
    }
};